week3/homework/src/index.js

```diff
@@ -1,3 +1,119 @@
 'use strict';
-
 // TODO: Write the homework code in this file
+const express = require('express');
+const uuid = require('uuid/v4');
+const app = express();
+const { readFile: _readFile, writeFile: _writeFile } = require('fs');
+const { promisify } = require('util');
+
+const readFile = promisify(_readFile);
+const writeFile = promisify(_writeFile);
+//build-in json middleware to automatically parse json
+app.use(express.json());
+
+const TODO_FILE = 'todos.json';
+
+function readTodos() {
+  return readFile(TODO_FILE, `utf8`).then(
+    JSON.parse,
+    () => ({})
+  );
+}
+
+function writeTodos(data) {
+  return writeFile(TODO_FILE, JSON.stringify(data));
+}
+
+//Read all todos
+app.get('/todos', async (req, res) => {
+  const todos = await readTodos();
+  res.json(todos);
+});
+
+//Read a todo by ID
+app.get('/todos/:id', async (req, res) => {
+  const id = req.params.id;
+  const todos = await readTodos();
+  if (todos.hasOwnProperty(id)) {
+    res.json(todos[id]);
+  }
+  else {
+    res.json(`id not found`);
+  }
+});
+
+//Create a todo
+app.post('/todos', async (req, res) => {
+  const newTodo = req.body;
+  const id = uuid();
+  const todos = await readTodos();
+  todos[id] = newTodo;
+  todos[id].done = false;
+  await writeTodos(todos);
+  res.json(todos);
+});
+
+//Delete a todo by ID
+app.delete('/todos/:id', async (req, res) => {
+  const id = req.params.id;
+  const todos = await readTodos();
+  if (todos.hasOwnProperty(id)) {
+    delete todos[id];
+    await writeTodos(todos);
+  }
+  else {
+    res.json(`id not found`);
+  }
+});
+
+//Delete all todos
+app.delete('/todos', async (req, res) => {
+  await writeTodos({});
+  res.json({});
+});
+
+//update a todo
+app.put('/todos/:id', async (req, res) => {
+  const id = req.params.id;
+  const todos = await readTodos();
+  const updatedTodo = req.body;
+  if (todos.hasOwnProperty(id)) {
+    todos[id].description = updatedTodo.description;
```

Comment:
@remarcmij Pointed to me that the way you are storing the todos doesn't actually fit with the way you are updating them. You are getting something like this from the requests:
```
{
  "todo": {
    "description": "(todo description)"
  }
}
```
Since you store the body of the request directly, your `JSON` file looks like this after adding a todo:
```
{
    "4cdcf95a-37e9-4d0f-aa06-4ed9f5818b57": {
        "todo": {
            "description": "(todo description)"
        },
        "done": false
    }
} 
```
Then if you update it like you're doing it now, it looks like this: 
```
{
    "4cdcf95a-37e9-4d0f-aa06-4ed9f5818b57": {
        "todo": {
            "description": "(todo description)"
        },
        "description": "updated"
        "done": false
    }
} 
```
I missed it at first, sorry about that :S

Reply:
I do every update to each todos[id] so, it will make the update of description on todos[id].description only and I try it on my lap and show me another output I can send you a screenshot of it

Reply:
Hi Hadeel, if I may chip in. You are right that your application works fine when you post TODOs like you do. However, the assignment gave a specific format for the request body format:

```json
{
  "todo": {
    "description": "(todo description)"
  }
}
```

See this link:

https://github.com/HackYourFuture/Node.js/tree/master/week3#request-body-format

Therefore your implementation does not comply with the specification. This will give issues if people attempt to use your backend based on the specification.